resourceIndexer/coreStagingS3.js

```diff
@@ -0,0 +1,292 @@
+import {logger} from './logger.js';
+import { SOURCE, VALID_AUSPICE_PATTERNS, INVALID_AUSPICE_PATTERNS,
+  DATESTAMP_REGEX, MAIN_DATASET_JSON } from './constants.js';
+
+/**
+ * The inventory of buckets (especially the core bucket) is in some ways a
+ * historical record of work over the years, but this isn't really what we want
+ * to display to users. As some examples:
+ *    - Files which don't match a resource to list should be excluded
+ *    - Datestampted files (i.e. _YYYY-MM-DD in the filename) are excluded
+ *      (we use S3 versioning instead)
+ * 
+ * If the s3 object is to be excluded we return false here.
+ * 
+ * In the case where the object represents a (part of) a resource we want to
+ * expose, then we categorise it here by adding the following properties:
+ *   - source (STAGING or CORE)
+ *   - resourceType (dataset, narrative or intermediate)
+ *   - id (the ID by which objects will be grouped together.
+ *        For datasets this is the nextstrain.org URL path, without any temporal signifier)
+ *   - subresourceType (currently only for resourceType=dataset)
+ */
+function categoriseCoreObjects(item, staging) {
+  const key = item.key
+  item.source = staging ? SOURCE.STAGING : SOURCE.CORE;
+  item.baseUrl = `https://nextstrain-${staging?'staging':'data'}.s3.amazonaws.com/${key}`
+  if (key.startsWith('search_')
+    || key.startsWith('manifest_')
+    || key.startsWith('datasets_')
+  ) return false;
+
+  // On the core bucket, directory-like hierarchies are used for intermediate
+  // files only (not auspice files). There are other resourceTypes here we may
+  // consider in the future -- e.g. model output JSONs
+  if (key.includes("/")) {
+    if (staging===true) return false;
+    if (key.startsWith('files/')) {
+      if (
+        key.includes('/archive/')
+        || key.includes('/test/')
+        || key.includes('/branch/')
+        || key.includes('/trial/')
+        || key.includes('/test-data/')
+        || key.includes('jen_test/')
+        || key.match(/\/\d{4}-\d{2}-\d{2}_results.json/) // forecasts-ncov
+        || key.endsWith('.png')                          // forecasts-ncov
+      ) {
+        return false;
+      }
+      item.resourceType = 'intermediate';
```

Comment:
I don't think I use the word "nonsensical" the same way you do. Here do you mean that we shouldn't display these to the user? If that's the case, I agree, and I've added an appropriate exclusion pattern to this block.

Reply:
I mean that it doesn't seem to make any sense to group together an intermediate like this:

    files/workflows/monkeypox/translations.zip

with one like this:

    files/workflows/forecasts-ncov/gisaid/nextstrain_clades/global.tsv.gz

solely because they were updated on the same date (`2023-10-26`).